site/content/rfds/0004-plugin-api.md

```diff
@@ -0,0 +1,932 @@
+  consumes in the course of its own execution.
+  - (Child) __Plugin__ (`Node`): An individual plugin node, specifying the
+    following attributes:
+    - (Positional Attribute) __Name__ (`String`): The name of the plugin, of
+      the form `<PUBLISHER>/<NAME>`.
+    - (Named Attribute) __Version__ (`String`): The version, specified as a
+      SemVer version constraint of [the forms accepted by Rust's Cargo
+      ecosystem][cargo_semver]. Namely, the following operators are supported
+      (note that this list explicitly _does not_ include
+      [hyphen-ranges][npm_hyphen_ranges] or [x-ranges][npm_x_ranges], both of
+      which are supported in the NPM ecosystem.)
+      - _Caret Requirements_: Permits "SemVer-compatible" updates. Example:
+        `^1.2.3`.
+      - _Tilde Requirements_: Specifies a minimal version with some ability to
+        update. Example: `~1.2.3`.
+      - _Wildcard Requirements_: Permits any version in the place of the
+        wildcard. Example: `1.2.*`
+      - _Comparison Requirements_: Manually specify a version constraint using
+        less than, less than or equal to, equal to, greater than or equal to,
+        or greater than operators. Example: `>= 1.2.3`.
+      - _Multiple Version Requirements_: Permits combining any of the above
+        forms of constraints by separating them with a comma.
+        Example: `>= 1.2.3, < 2.0.0`.
+    - (Named Attribute) __Manifest__ (Optional / `String`): The download
+      location of the __plugin discovery manifest__ for this plugin. Until a
+      plugin registry is established for Hipcheck plugins, this attribute will
+      effectively be required, as otherwise Hipcheck will have no mechanism to
+      discover the location of the download metadata for a plugin.
+
+Given the above definitions, an example plugin manifest may look like:
+
+```kdl
+publisher "mitre"
+name "affiliation"
+version "0.1.0"
+license "Apache-2.0"
+entrypoint {
+  on arch="aarch64-apple-darwin" "./hc-mitre-affiliation"
+  on arch="x86_64-apple-darwin" "./hc-mitre-affiliation"
+  on arch="x86_64-unknown-linux-gnu" "./hc-mitre-affiliation"
+  on arch="x86_64-pc-windows-msvc" "./hc-mitre-affiliation"
+}
+
+dependencies {
+  plugin "mitre/git" version="0.1.0" manifest="https://github.com/mitre/hipcheck/blob/main/plugin/dist/mitre-git.kdl"
+}
+```
+
+The __plugin discovery manifest__ is an additional manifest which tells
+Hipcheck where to find the correct plugin artifact for a specific target
+architecture. In the future, Hipcheck's plugin system may support one or
+more plugin registries which ease discovery of Hipcheck plugins. In the
+short term, the standard mechanism for discovery of Hipcheck plugins
+will be for plugin producers to distribute a per-plugin discovery manifest
+which contains version and target architecture information necessary for
+Hipcheck to download the _correct_ plugin executable artifact to run
+on the end-user's host.
```

Comment:
Do we want to consider at this time the possibility of some users wanting to run in a fully-offline mode with plugins they already have on-disk?

Reply:
Good point! I think this would properly go in the "Running Plugins" section.